quickwit-metastore/src/metastore/postgresql_metastore.rs

```diff
@@ -732,6 +734,187 @@ impl Metastore for PostgresqlMetastore {
+    async fn create_delete_task(&self, delete_query: DeleteQuery) -> MetastoreResult<DeleteTask> {
+        run_with_tx!(self.connection_pool, tx, {
+            // Serialize the delete query to fit the database model.
+            let delete_query_json = serde_json::to_string(&delete_query).map_err(|err| {
+                MetastoreError::InternalError {
+                    message: "Failed to serialize delete query.".to_string(),
+                    cause: err.to_string(),
+                }
+            })?;
+            // Create delete task.
+            let (create_timestamp, opstamp): (sqlx::types::time::PrimitiveDateTime, i64) =
+                sqlx::query(
+                    "INSERT INTO delete_tasks (index_id, delete_query_json) VALUES ($1, $2) \
+                     RETURNING create_timestamp, opstamp",
+                )
+                .bind(&delete_query.index_id)
+                .bind(&delete_query_json)
+                .fetch_one(tx)
+                .await
+                .map(|row| (row.get(0), row.get(1)))
+                .map_err(|err| convert_sqlx_err(&delete_query.index_id, err))?;
+            Ok(DeleteTask {
+                create_timestamp: create_timestamp.assume_utc().unix_timestamp(),
+                opstamp: opstamp as u64,
+                delete_query,
+            })
+        })
+    }
+
+    /// Deletes delete tasks of a given index.
+    #[instrument(skip(self))]
+    async fn delete_delete_tasks(&self, index_id: &str) -> MetastoreResult<()> {
+        run_with_tx!(self.connection_pool, tx, {
+            sqlx::query("DELETE FROM delete_tasks WHERE index_id = $1")
+                .bind(index_id)
+                .execute(tx)
+                .await?;
+            Ok(())
+        })
+    }
+
+    /// Update splits delete opstamps.
+    #[instrument(skip(self))]
+    async fn update_splits_delete_opstamp<'a>(
+        &self,
+        index_id: &str,
+        split_ids: &[&'a str],
+        delete_opstamp: u64,
+    ) -> MetastoreResult<()> {
+        if split_ids.is_empty() {
+            return Ok(());
+        }
+        run_with_tx!(self.connection_pool, tx, {
+            let updated_split_ids: Vec<String> = sqlx::query(
+                r#"
+                UPDATE splits
+                SET delete_opstamp = $1
+                WHERE
+                    index_id = $2
+                    AND split_id = ANY($3)
+                    AND split_state = $4
+                RETURNING split_id
+            "#,
+            )
+            .bind(delete_opstamp as i64)
+            .bind(index_id)
+            .bind(split_ids)
+            .bind(SplitState::Published.as_str())
+            .map(|row| row.get(0))
+            .fetch_all(&mut *tx)
+            .await?;
+            // If no splits were returned, maybe the index itself does not exist
+            // in the first place.
+            if updated_split_ids.is_empty() && index_opt(tx, index_id).await?.is_none() {
```

Comment:
nitpick....

Using RETURNING and fetch the split_ids was overkill here.
You only use their count.

The following function gives you the number of modified rows

`sqlx_result.rows_affected()`
https://docs.rs/sqlx/latest/sqlx/postgres/struct.PgQueryResult.html#method.rows_affected